config/src/main/java/org/springframework/security/config/annotation/authentication/configuration/AuthenticationConfiguration.java

```diff
@@ -151,10 +152,23 @@ public void setObjectPostProcessor(ObjectPostProcessor<Object> objectPostProcess
 		if (beanNamesForType.length == 0) {
 			return null;
 		}
-		Assert.isTrue(beanNamesForType.length == 1,
-				() -> "Expecting to only find a single bean for type " + interfaceName
-						+ ", but found " + Arrays.asList(beanNamesForType));
-		lazyTargetSource.setTargetBeanName(beanNamesForType[0]);
+		String beanName;
+		if (beanNamesForType.length > 1) {
+			List<String> primaryBeanNames = Arrays.stream(beanNamesForType)
```

Comment:
This should be able to be simplified to something like:

```java
applicationContext.getAutowireCapableBeanFactory().resolveNamedBean(Foo.class);
```

Reply:
Hi @rwinch, would that replace whole `BeanFactoryUtils.beanNamesForTypeIncludingAncestors(applicationContext, interfaceName);` part? If I understand correctly, since the implementation of `resolveNamedBean` handles primary, this method could just become something along the lines of

```java
LazyInitTargetSource lazyTargetSource = new LazyInitTargetSource();
String beanName;
try {
	beanName = applicationContext.getAutowireCapableBeanFactory().resolveNamedBean(interfaceName).getBeanName();
} catch (NoSuchBeanDefinitionException ex) {
	return null;
}
lazyTargetSource.setTargetBeanName(beanName);
lazyTargetSource.setBeanFactory(applicationContext);
ProxyFactoryBean proxyFactory = new ProxyFactoryBean();
proxyFactory = objectPostProcessor.postProcess(proxyFactory);
proxyFactory.setTargetSource(lazyTargetSource);
return (T) proxyFactory.getObject();
```

However, is this method still lazy if we do that? It looks to me like `resolveNamedBean` will init an instance of the bean.

Reply:
Thanks for pointing that out. I missed that it did create the bean too.